### ISSUE TITLE
Monospace font rendering issues in Alacritty with 2.2.1 but not 2.1.0

### ISSUE THREAD

**Issue by liamtimms** [ryanoasis/nerd-fonts#896]
#### 🗹 Requirements

- [X] I have searched the [issues](https://github.com/ryanoasis/nerd-fonts/issues) for my issue and found nothing related and/or helpful
- [X] I have searched the [FAQ](https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting) for help
- [X] I have searched the [Wiki](https://github.com/ryanoasis/nerd-fonts/wiki) for help

#### 🎯 Subject of the issue

Experienced behavior:

Major differences in font rendering for jetbrains between releases `2.1.0` and `2.2.1`.

Expected behavior:

I'd expect that spacing/rendering would continue working as in previous versions. 

Example symbols:

**left** window is with release `2.1.0` while **right** is `2.2.1`: 

![image](https://user-images.githubusercontent.com/14281621/187034258-bb835d3a-e872-420d-8eeb-3acb62a577cf.png)

#### 🔧 Your Setup

- _Which font are you using (e.g. `Anonymice Powerline Nerd Font Complete.ttf`)?_
   - _Please give the full filename_
   - _Where did you get the file from (download link, self patched, source downloaded from link...)_
- _Which terminal emulator are you using (e.g. `iterm2`, `urxvt`, `gnome`, `konsole`)?_
- _Are you using OS X, Linux or Windows? And which specific version or distribution?_

**font:**  ` family: jetbrainsmono medium nerd font `
**from:** [nerd-fonts-jetbrains-mono](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nerd-fonts-jetbrains-mono) via AUR
**terminal:** alacritty
**OS:** Arch Linux

#### ★ Screenshots (Optional)

**left** window is with release `2.1.0` while **right** is `2.2.1`: 

![image](https://user-images.githubusercontent.com/14281621/187034170-bda2eff8-a8a0-4c11-95ad-1e1f8def51cb.png)

Also of note: 

![image](https://user-images.githubusercontent.com/14281621/187034568-7c44bc3c-52ec-4954-9df0-9766dd313e73.png)

The install size appears to have changed significantly between the versions.

**Comment by Finii** [comment]
Sorry to hear of and see your problem with 2.2.1.

The right image looks like it does mix fonts that should not be used together.

The JetBrains got another variant now, `JetBrains Mono Nerd Font` and `JetBrains Mono NL Nerd Font`:
https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/JetBrainsMono

The individual font files look fine at first glance.
The font family naming is buggy, but that has already been so in 2.1.0

Some questions that would help reproducing
* Is the effect also visible in `gnome-terminal`, or `tilix`, or `kitty` (spares me learning `alacritty`)?
* Did you reboot (I know usually not needed)?
* Can you share the `alacritty` configuration (screenshot or something)?
* Do you have the 'regular' Jet Brains Mono installed in parallel?
* Would you have time and like to test a new set of `JetBrainsMono NF` when I prepare one?

This AUR installs 128 font files. With the broken family naming this can be problematic.

**Comment by simonwiles** [comment]
I had the same issue with `konsole` (`nerd-fonts-jetbrains-mono` installed from the AUR on Arch). Fwiw the `PKGBUILD` from the AUR installs 64 fonts for me, including the new NL variant (which is very welcome, thanks!).

Noodling around rolling back to `2.1.0` and then reinstalling `2.2.1` for testing purposes, I am no longer able to reproduce the issue.  I think it was caused by a faulty post-install script that the AUR maintainer has now removed (as of 2 hours ago, commit `ab7a714f`).

**Comment by liamtimms** [comment]
@simonwiles I just tested with the new commit and the issue is still present after a reboot. 

@Finii can you explain the "font family naming is buggy" a little bit more? 


> Is the effect also visible in gnome-terminal, or tilix, or kitty (spares me learning alacritty)?

It seems to be ok in gnome-terminal:

![image](https://user-images.githubusercontent.com/14281621/187047345-8b548fbc-8c59-458a-bffb-ab73f2cbdea0.png)

Yet, gnome-terminal also doesn't let me pick the medium variant, which I have found much more readable. I know that Alacritty does not support ligatures and now see that "NL" stands for "no ligatures"; if ligatures handling has changed that might be part of the issue, but specifying `jetbrainsmononl` seems not actually to be affecting it. Possibly this is due to the buggy behavior around picking the font family you mentioned. I can file something on the alacritty bug tracker if that's more appropriate but I am keeping the nerd front downgraded in the meantime. 

> Did you reboot (I know usually not needed)?

Yep. 

> Can you share the alacritty configuration (screenshot or something)?

<details><summary>click for font config</summary>
<p>

```yaml
# Font configuration
font:
  # Normal (roman) font face
  normal:
    # Font family
    #
    # Default:
    #   - (macOS) Menlo
    #   - (Linux/BSD) monospace
    #   - (Windows) Consolas
    #
    #family: monospace
    # family: source code pro semibold
    family: jetbrainsmono medium nerd font 

    # The `style` can be specified to pick a specific face.
    #style: Regular

  # Bold font face
  bold:
    # Font family
    #
    family: jetbrainsmono Extrabold nerd font
    # If the bold family is not specified, it will fall back to the
    # value specified for the normal font.
    # family: source code pro

    # The `style` can be specified to pick a specific face.
    style: Extrabold

  # Italic font face
  italic:
    # Font family
    #
    # If the italic family is not specified, it will fall back to the
    # value specified for the normal font.
    # family: source code pro semibold

    # The `style` can be specified to pick a specific face.
    style: Italic

  # Bold italic font face
  bold_italic:
    # Font family
    family: jetbrainsmono Extrabold Italic nerd font
    #
    # If the bold italic family is not specified, it will fall back to the
    # value specified for the normal font.
    # family: monospace

    # The `style` can be specified to pick a specific face.
    style: Bold Italic

  # Point size
  size: 12.5
```

</p>
</details>


> Do you have the 'regular' Jet Brains Mono installed in parallel?

Nope, only the nerd font versions. 


> Would you have time and like to test a new set of JetBrainsMono NF when I prepare one?`

Sure. 

> This AUR installs 128 font files. With the broken family naming this can be problematic.

Should the NL fonts be separated into their own package?

**Comment by liamtimms** [comment]
As a quick followup: switching from medium weight font improves but does not resolve the issue

with medium ( `family: jetbrainsmono medium nerd font`)


![image](https://user-images.githubusercontent.com/14281621/187047814-1cc32444-b558-4d3d-8b51-5b1c833f7780.png)

without medium (`family: jetbrainsmono nerd font`)


![image](https://user-images.githubusercontent.com/14281621/187047795-d89257d2-f01b-4582-bce6-da70d53c4ab0.png)

**Comment by Finii** [comment]
> with medium ( family: jetbrainsmono medium nerd font)

Hmm, the Family name for the `JetBrainsMono Nerd Font Mono` ... is that not different?

See here..

![image](https://user-images.githubusercontent.com/16012374/187048111-713f732b-e287-48a9-88ef-f5f688888da0.png)

**Comment by liamtimms** [comment]
Honestly, I have no idea. I originally tried different combinations until I got something that worked in the configuration with previous releases. I mention it mainly as a clue since it is slightly better without `medium` specified but still borked. While specifying the family like that seems odd, it was working without issue in 2.1.0 and does work if I downgrade back to 2.1.0. It's possible that some change in 2.2.1 has exposed an underlying bug in font selection, rendering, ligature handling, or some combination thereof.

**Comment by liamtimms** [comment]
Wait ok, I think it is something to do with font family selection. Switching the family to `jebrainsmono nerd font` and then specifying medium style *and* dropping family selection from the Bold specification seems to work. 

![image](https://user-images.githubusercontent.com/14281621/187048678-2715a722-bdaa-4e2b-8a93-31e31508ba0b.png)

This seems like the intended way of configuring the font, and _I think it's actually that something in font family selection was broken previously, and my workaround had stopped working with this update._

**Comment by Finii** [comment]
> Should the NL fonts be separated into their own package?

The same question holds for:
* Should the 'Windows Compatible' fonts not be separately packed?
* Why are 'Windows Compatible' fonts in an AUR package anyway?
* Does it make sense to install both 'Nerd Font' and 'Nerd Font Mono' variants, or should they be separated?

At the moment we have no separation at all. For example _Terminus_ has not only two variants like JetBrains (i.e. NL and 'with ligatures'), but 'BDF', 'PCF', 'PSF', and '4.40' :grimacing: 

_WHAT? Just checked the Terminus release package, and there is only one of the 4 variants?_

Hmm, 'Gohu' has also 4 variants, ... but these are packed at least in subdirs?

OK, I will check that.

> @Finii can you explain the "font family naming is buggy" a little bit more?

The patched font naming is ... strange, due to historical reasons. A lot of patched fonts just get clearly wrong names.

For example have a look here: https://github.com/ryanoasis/nerd-fonts/blob/de9bb5b9bff7052478ebf059115a977d5df34b95/bin/scripts/name_parser/name_parser_test2.known_issues#L313-L316

The lines in that file are:
 `>`: Unpatched (original) font name
`+`: Corrected family naming
`=`: Current family naming
Notable differences are marked with an 'X'.

The font is _Iosevka Term Bold Oblique_.
But 'Term' is wrongly dropped from the naming, thus mixing Iosevka (normal) and Iosevka Term :-(

With 2.2.0 `font-patcher` got a new parameter `--makegroups` that will correct that. At the moment it is only used for Cascadia Code.

See also #717, which has been changed afterwards and is introduced via #723.

> Possibly this is due to the buggy behavior around picking the font family you mentioned. I can file something on the alacritty bug tracker if that's more appropriate but I am keeping the nerd front downgraded in the meantime.

Let me check some things first. Just installed all fonts and starting an investigation. Need to get alacritty... :hammer_and_wrench: 

Ah, here, after installing all fonts in the AUR:
![image](https://user-images.githubusercontent.com/16012374/187067542-ed6cd755-a13c-4336-9b40-1d6198c16343.png)

`NF` is the Windows Compatible version, and you see there is no `NF Mono`, because the family name is broken and with windows compatible mono and non-mono fonts mix :-(

> > Did you reboot (I know usually not needed)?
> 
> Yep.

Sorry I had to ask the obvious ;)

> > Would you have time and like to test a new set of JetBrainsMono NF when I prepare one?`
> 
> Sure.

Let me create an experimental package with corrected Family names. I will put the link **here** when it is ready (need to feed my Kids first).

> I think it's actually that something in font family selection was broken previously, and my workaround had stopped working with this update

Hmm. Lets see, with all the AUR supplied fonts installed, I get this

```shell
$ fc-list : family | grep Jet
JetBrainsMonoNL NF
JetBrainsMono Nerd Font
JetBrainsMonoNL Nerd Font Mono
JetBrainsMonoNL Nerd Font
JetBrainsMono Nerd Font Mono
JetBrainsMono NF
```

Then lets assume you want `JetBrainsMonoNL` (non-Mono), lets see which styles are there

```shell
$ fc-list : family style | grep "JetBrainsMonoNL Nerd Font:"
JetBrainsMonoNL Nerd Font:style=Medium Italic,Italic
JetBrainsMonoNL Nerd Font:style=ExtraLight Italic,Italic
JetBrainsMonoNL Nerd Font:style=Light Italic,Italic
JetBrainsMonoNL Nerd Font:style=ExtraBold Italic,Italic
JetBrainsMonoNL Nerd Font:style=Bold
JetBrainsMonoNL Nerd Font:style=ExtraLight,Regular
JetBrainsMonoNL Nerd Font:style=Light,Regular
JetBrainsMonoNL Nerd Font:style=SemiBold Italic,Italic
JetBrainsMonoNL Nerd Font:style=Medium,Regular
JetBrainsMonoNL Nerd Font:style=SemiBold,Regular
JetBrainsMonoNL Nerd Font:style=Italic
JetBrainsMonoNL Nerd Font:style=Thin,Regular
JetBrainsMonoNL Nerd Font:style=ExtraBold,Regular
JetBrainsMonoNL Nerd Font:style=Bold Italic
JetBrainsMonoNL Nerd Font:style=Regular
JetBrainsMonoNL Nerd Font:style=Thin Italic,Italic
```

Ok, and you want `Medium`...

```shell
$ fc-list : family style | grep "JetBrainsMonoNL Nerd Font:" | grep Medium
JetBrainsMonoNL Nerd Font:style=Medium Italic,Italic
JetBrainsMonoNL Nerd Font:style=Medium,Regular
```

Of course `Medium` has no `Bold` style, just `Regular` and `Italic`, so you need to provide a different font for the bold console stuff (probably).

**Comment by Finii** [comment]
Hmm, the `v2.1.0` has very strange filenames...

Just focusing on the 'normal' variants, i.e. not `Nerd Font Mono` and not `Windows Compat`.

![image](https://user-images.githubusercontent.com/16012374/187082048-5f5c87a8-466b-4abf-b34b-384830965ff1.png)

`JetBrains Mono Medium` **and** `JetBrains Mono Medium Medium`?
And what is `JetBrains Mono Medium Med Ita` compared to `JetBrains Mono Medium Italic`?
`ExtraBold ExtBd` and `ExtraBold ExBd` and `Extra Bold`?

And this is just the filenames. The fontconfig sees this:

![image](https://user-images.githubusercontent.com/16012374/187082468-1b73b2f1-eb61-4a45-b16d-db6576cfce77.png)

Ah I get it, somehow there was a last-minute rebuilt of the patched fonts, but the older version were not removed :roll_eyes: 

![image](https://user-images.githubusercontent.com/16012374/187082663-ea00fbe2-06f8-4d4c-a8c0-63d5bb730e6d.png)

I will try to compile a history of commits that change the family naming behavior ...

**Comment by Finii** [comment]
As example we follow `JetBrains Mono Medium`:

* #421 (8c54efe61b3c) `JetBrains Mono Medium Nerd Font Complete.ttf`
* dd34cef6c9a (Update to 1.0.2) `JetBrains Mono Medium Medium Nerd Font Complete.ttf`
* Release `v2.1.0` (fa69496)
* 101fc6e3581184f862e353da2b3780394a193bb5 (Update to 1.0.5) `JetBrains Mono Medium Nerd Font Complete.ttf`
* #518 adds 'no ligature' variant
* #540
* #572 
* #648 
* #776
* Release `v2.2.0`
* Release `v2.2.1`

Hmm, and then I found [this](https://github.com/ryanoasis/nerd-fonts/commit/fa69496f5a3291fbcb6c0ad5bcec9c8bcab842fc), the comments are from January 2022:

![image](https://user-images.githubusercontent.com/16012374/187083916-f88d9e17-8f72-4964-9f3c-72646ccbbb5d.png)

Interestingly the family naming in the `Medium Medium` font is correct, while in the other it is not.
And the reason is not a change in our `font-patcher`, but in the source fonts (the original unpatched ones), note the peculiar naming in 1.0.2:

![image](https://user-images.githubusercontent.com/16012374/187085763-4b5634cc-07c8-4a74-8396-2e5211f5ad56.png)

So the one working for you was the one buggy version that noone cleaned out ;)

**Comment by liamtimms** [comment]
> So the one working for you was the one buggy version that noone cleaned out ;)

Great, that makes sense and was my intuition from the fact that I only got it working originally through trial + error while the new configuration which is working seems much more sensible. This is a very interesting journey and also explains why I was having difficulty selecting the older version of the font in other programs.

**Comment by Finii** [comment]
If you like, you can find a patched JetBrainsMono font with 'correct' Family naming here:

https://github.com/Finii/nerd-fonts/tree/experimental/JetBrainsMono-makegroups/archives

It uses the new `--makegroups` flag. Would be good to test this, as it is a candidate for Nerd Fonts 3.0.0 patched font naming :-)

Note that also the file names changed to a more standard-ish format with the style last. They still have the unexpected blanks in them, ...

**Comment by beyond9thousand** [comment]
Could this be related? Its been bothering me like crazy since my fresh linux install

![image](https://user-images.githubusercontent.com/77363063/194730087-70087cc4-860d-4560-9a7a-4cdfd47483fe.png)

**Comment by Finii** [comment]
@beyond9thousand 
> Could this be related? Its been bothering me like crazy since my fresh linux install

Not good if it drives you crazy.

Which font did you select on https://www.programmingfonts.org, I can not see that in the screenshot?
I guess you type `>` in that browser window, and copy and paste that into the Alacritty window?
Can you copy & paste that also in a text comment here, or give the unicode of that `>` like thing?

Maybe good would be the relevant Alacritty settings, and the output of `ls -l` for the font file you use.

**Comment by beyond9thousand** [comment]
Apologies, for further context - the character you see is actually the zsh prompt for my shell. I copied it from my terminal and pasted it over to https://www.programmingfonts.org/#jetbrainsmono

CHAR: `❯`
UNICODE: `U+276F`

`alacritty.yml`:
```
font:
  normal:
    family: "JetBrainsMono Nerd Font"
```

Font files in use:
```-rw-r--r-- 1 root root 1094232 Sep 22 04:36 JetBrains Mono Bold Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1115436 Sep 22 04:36 JetBrains Mono Bold Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1089884 Sep 22 04:36 JetBrains Mono Bold Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1111088 Sep 22 04:36 JetBrains Mono Bold Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1095412 Sep 22 04:36 JetBrains Mono ExtraBold Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1116616 Sep 22 04:36 JetBrains Mono ExtraBold Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1091412 Sep 22 04:36 JetBrains Mono ExtraBold Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1112616 Sep 22 04:36 JetBrains Mono ExtraBold Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1089256 Sep 22 04:36 JetBrains Mono ExtraLight Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1110460 Sep 22 04:36 JetBrains Mono ExtraLight Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1086564 Sep 22 04:36 JetBrains Mono ExtraLight Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1107768 Sep 22 04:36 JetBrains Mono ExtraLight Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1091224 Sep 22 04:36 JetBrains Mono Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1112428 Sep 22 04:36 JetBrains Mono Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1091592 Sep 22 04:36 JetBrains Mono Light Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1112796 Sep 22 04:36 JetBrains Mono Light Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1088476 Sep 22 04:36 JetBrains Mono Light Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1109680 Sep 22 04:36 JetBrains Mono Light Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1091272 Sep 22 04:36 JetBrains Mono Medium Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1112476 Sep 22 04:36 JetBrains Mono Medium Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1086824 Sep 22 04:36 JetBrains Mono Medium Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1108028 Sep 22 04:36 JetBrains Mono Medium Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1086628 Sep 22 04:36 JetBrains Mono Regular Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1107836 Sep 22 04:36 JetBrains Mono Regular Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1094672 Sep 22 04:36 JetBrains Mono SemiBold Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1115876 Sep 22 04:36 JetBrains Mono SemiBold Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1089616 Sep 22 04:36 JetBrains Mono SemiBold Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1110820 Sep 22 04:36 JetBrains Mono SemiBold Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1088500 Sep 22 04:36 JetBrains Mono Thin Italic Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1109704 Sep 22 04:36 JetBrains Mono Thin Italic Nerd Font Complete.ttf
-rw-r--r-- 1 root root 1084320 Sep 22 04:36 JetBrains Mono Thin Nerd Font Complete Mono.ttf
-rw-r--r-- 1 root root 1105524 Sep 22 04:36 JetBrains Mono Thin Nerd Font Complete.ttf
```

**Comment by Finii** [comment]
Ah, thank you.
`276F` is not patched in by Nerd Fonts, so either it is in the source or not.

It **is** in the source in the currently used source font:

![image](https://user-images.githubusercontent.com/16012374/194753837-285fa7df-5d64-4c48-8118-c666bdea1386.png)

and it is in the Release (just one font checked though):

![image](https://user-images.githubusercontent.com/16012374/194754105-e935454b-08c2-4fee-8f15-b28ca8c5c40a.png)

And that glyph looks exactly like the one in your Alacritty screenshot.
Seems the `programmingfonts.org` uses an old JetBrains Mono? That codepoint has been added somewhen to JB-Mono, iirc.

**Comment by beyond9thousand** [comment]
Odd. Before my current linux install, I was using the exact same font and the same alacritty config along with exactly the same zsh prompt, and it looked like it does on the `programmingfonts.org`

The JB-Mono that you mention, is it a separate font file?

**Comment by Finii** [comment]
Interestingly the github repo of programmingfonts.org has a JetBrainsMono with that same glyph:

![image](https://user-images.githubusercontent.com/16012374/194754669-e796bc77-c585-4f25-bc51-ace230f854c4.png)

At least the `ttf`, did not check the VF etc

> The JB-Mono that you mention, is it a separate font file?
I do not underdstand? I used 'JB-Mono' just to not type 'JetBrains Mono'.

**Comment by beyond9thousand** [comment]
My bad, I thought you were suggesting that a separate version of the font file exists with the glyph that I was accustomed to up until recently. Also, I just confirmed your suspicion.

It was indeed swapped out/removed in one of the new releases. I installed an outdated version of JB Nerd Fonts from the aur (the package itself is marked outdated)

I am currently on `2.1.0-1` and it renders the glyph as I remember it to be
![image](https://user-images.githubusercontent.com/77363063/194754923-1d6db5c3-227d-436a-b14e-2a7fca762731.png)

**Comment by Finii** [comment]
Hmm, here I typed `>` (ordinary bigger-then), `u276f` and `u2771`:

![image](https://user-images.githubusercontent.com/16012374/194754994-3f18c65e-b51f-450a-8384-1f68a89188e9.png)

Strange is that the two `27xx` chars seem to have different height, while they should not according to the font itself:

![image](https://user-images.githubusercontent.com/16012374/194755042-fb87059f-e60f-40cb-906c-b65ad732cfb2.png)

**Comment by Finii** [comment]
I'm not really sure now ... is it 'solved' for you our shall we try something still?

**Comment by beyond9thousand** [comment]
It is solved so to speak, but the discrepancies you point out are bizarre nonetheless...

**Comment by Finii** [comment]
> bizarre

Well, if you turn the font on and off on the web page, notice that some `}` change and others not? Also the text in the top compared to in the middle?!

![jbm](https://user-images.githubusercontent.com/16012374/194755526-4bcd778b-5234-4fe2-bac1-c508e579c54c.gif)

**Comment by beyond9thousand** [comment]
I am having a r/thanksihateit episode

**Comment by Finii** [comment]
Heavy Brackets now patched in via commit 59c5cb5a1f33139b79838f010fafea26623ed046  `font-patcher: Patch in heavy angle brackets`

**Comment by beyond9thousand** [comment]
Seems that further maintenance on that front is not guaranteed from the changelog, unless I am interpreting it incorrectly
